### PR TITLE
feat: add provd identity service

### PR DESCRIPTION
**Pull request by d-loose**
Replaces the DBus-based XDG identity service with one based on provd.

Since the backend doesn't maintain any state, the configured identity is saved in the service so that it can be retrieved later to launch the user's desktop session (via #298).

~~Some things that need to be figured out:~~

- ~~We probably don't want to create a user account immediately after the user presses 'next' on the identity page. Either we have to maintain a state in provd like subiquity does (in that case we should use similar methods, e.g. `setIdentity` instead of `createUser` and add an `apply` method that carries out the user creation) , or we need to handle this in the UI and keep a state in the client.~~
- ~~Depending on how we want to handle the previous point, we might not need a `getUser` method in provd. It's supposed to return the ephemeral user account information configured in the UI before the account is created.~~

### CODE REVIEW

File: packages/provd_client/lib/src/provd_user_client.dart [hunk 1]
```diff
@@ -38,8 +39,12 @@ class ProvdUserClient {
...
   /// Validates the [username] and returns true if it is valid.
   Future<bool> validateUsername(String username) async {
-    final request = ValidateUsernameRequest(username: username);
-    final response = await _userClient.validateUsername(request);
-    return response.valid;
+    try {
+      final request = ValidateUsernameRequest(username: username);
+      final response = await _userClient.validateUsername(request);
+      return response.valid;
+    } on GrpcError catch (_) {
```
Comment (d-loose) [inline packages/provd_client/lib/src/provd_user_client.dart:46]: So far provd only returns a boolean here and propagates additional information (i.e. why the username is invalid) as grpc errors. We could either generalize the grpc response (similar to how subiquity does this) or handle the different grpc error cases (once defined) here in the client and then wrap the result in an enum again.

File: packages/ubuntu_init/lib/src/services/provd_identity_service.dart [hunk 2]
```diff
@@ -0,0 +1,41 @@
...
+
+  @override
+  Future<Identity> getIdentity() async => _identity;
+
+  @override
+  Future<void> setIdentity(Identity identity) async {
+    if (_identity == identity) return;
+    _identity = identity;
+    return apply();
+  }
```
Comment (d-loose) [inline packages/ubuntu_init/lib/src/services/provd_identity_service.dart:27]: See comment at the beginning. If we want to maintain a state here, we should call `apply` separately at a later point after which we can't return to the identity page in the UI.

**Comment by d-loose** [comment]
> We probably don't want to create a user account immediately after the user presses 'next' on the identity page. ...

Update: After some discussions about the first boot initialization step we could consider making the user creation irreversible, i.e. actually create the account when the user presses 'next' and remove the 'back' button on the following page.

Advanced users (especially those setting up a managed/enterprise machine) might want to open a terminal and perform actions using the new local user right away (while ubuntu-init is still running).

File: packages/provd_client/lib/src/provd_user_client.dart (hunk 1 above)
Comment (d-loose) [reply packages/provd_client/lib/src/provd_user_client.dart]: We decided to return an enum from the backend instead

File: packages/provd_client/lib/src/generated/accessibility.pb.dart
Comment (d-loose) [inline packages/provd_client/lib/src/generated/accessibility.pb.dart:1]: Oh, didn't mean to delete all those ... Will restore them

**Review by matthew-hagemann — CHANGES_REQUESTED** [review]
Hi Dennis,

I've run into a few issues with username validation in the new setup:

I am able to create a user from the reserved list (ie: root)
It looks like `provd` does fail, because if I click next, `ubuntu_init` gets stuck

Screenshot from 2024-01-18 12-27-47

Same issue for other reserved usernames. I also tried `gdm`

I also ran into an issue where valid usernames were rejected as being taken:

Screenshot from 2024-01-18 13-31-59

If it would be helpful, I can add some debug-logs to `provd` and push a new version up to help inspect.

**Comment by d-loose** [comment]
Do you have the logs from ubuntu_init? I can't reproduce it so far.

**Review by matthew-hagemann — APPROVED** [review]
LGTM!

**Comment by d-loose** [comment]
I assume all the issues you encountered were due to bugs in provd and/or your testing setup then?
In any case I'll only merge this once we're happy with the backend side :)

**Comment by matthew-hagemann** [comment]
@d-loose Yea, I can't reproduce any of my issues after redoing my test infrastructure :slightly_smiling_face:
Cool beans! I'll send a PR out in a bit, just making some last changes for Didier :+1:

**Pull request merged**